### DISCUSSION TITLE
How can I configure sync timeout?

### DISCUSSION THREAD

**Discussion by smeeklai** [argoproj/argo-cd#6208] [Q&A] [answered]
I deployed ArgoCD via Helm and I want to configure ArgoCD sync timeout because whenever there's an error occurred, ArgoCD retries and wait for a bit too long until I got sync failure notification.

**Comment by jgwest** [comment, marked as answer]
Hi @smeeklai,

There's no way to set a 'maximum time/expiration time' on the sync operation AFAICT, but you can configure the number of retries...

When creating an application (for example, via the CLI), the number of retries (and the time between retries) can be configured:
```
      --sync-policy string                         Set the sync policy (one of: none, automated (aliases of automated: auto, automatic))
      --sync-retry-backoff-duration duration       Sync retry backoff base duration. Input needs to be a duration (e.g. 2m, 1h) (default 5s)
      --sync-retry-backoff-factor int              Factor multiplies the base duration after each failed sync retry (default 2)
      --sync-retry-backoff-max-duration duration   Max sync retry backoff duration. Input needs to be a duration (e.g. 2m, 1h) (default 3m0s)
      --sync-retry-limit int                       Max number of allowed sync retries

```
(https://argoproj.github.io/argo-cd/user-guide/commands/argocd_app_create/)

Primarily for your question you would want to look at `--sync-retry-limit`.

Or you can add/tweak the same options on the Application CR:
https://github.com/argoproj/argo-cd/blob/ac2491d37dcfece786c69b0501af0a9be42f128d/docs/operator-manual/application.yaml#L111

For example:
```
    retry:
      limit: 1 # number of failed sync attempt retries; unlimited number of attempts if less than 0
```

**Reply by smeeklai** [reply]
Understood. Thank you :)

**Comment by benedikt-haug** [comment]
I would absolutely love a timeout for sync operations. Currently I have to manually track and cancel jobs that couldn't complete.
![image](https://user-images.githubusercontent.com/2534537/134643481-8b708ad3-e12d-4deb-922f-66be1d411ffd.png)

**Reply by melch** [reply]
We have a similar experience. Setting the max number of retries doesn't help our system because a single sync attempt can get stuck and never complete, so we won't hit a max number of retries.

**Reply by thapakazi** [reply]
![mr-bean-waiting](https://github.com/argoproj/argo-cd/assets/1695906/4d32425c-792d-4d57-887c-a3c958659b57)

**Comment by brat002** [comment]
Any progress or plans to include the feature into upcomming release? really annoying... :( Probably `retry` option doesn't work as expected.

**Comment by riuvshyn** [comment]
still no sync timeout???

**Reply by hellojanam** [reply]
we are facing this issue too, if we deployed something and that caused pods to crash then the sync never stops, which caused issue. we love to have sync time out.

**Comment by richardfelkl** [comment]
The sync timeout would be very much appreciated.

**Comment by erikschul** [comment]
Why is this a discussion instead of an issue?

Possibly related:
https://github.com/argoproj/argo-cd/issues/15642

**Reply by aslafy-z** [reply]
I think this is the original issue https://github.com/argoproj/argo-cd/issues/11494 and this the original discussion https://github.com/argoproj/argo-cd/discussions/11276

**Comment by riuvshyn** [comment]
Sync timeout is a must... there are too many things that might go wrong and result into a never ending sync that would last forever which atm require manual intervention. :/

**Comment by sudeepgupta90** [comment]
Using `argocd_app_info{sync_status="OutOfSync"}>=1` for 30 min to detect such scenarios. It doesn't fix this problem, but at least gives visibility from metrics side on Prometheus

**Comment by keskad** [comment]
In my case hanging Post-Sync jobs are causing e.g. 15 hours running synchronization. Timeout would be nice.

**Comment by szczepq** [comment]
We have a similar problem. One of ArgoCD is installed on a VM that we don't have access to so when we make a mistake and even fix it in the next commit we can't manually terminate the current ongoing sync and re-run the new one with the fix. ;/

**Comment by jeevanions** [comment]
This is big issues for us where we had to manually go through large number of ArgoCD apps and terminate the ongoing sync and re-sync. We would like some configurable parameters to play with around timeout.

**Reply by aslafy-z** [reply]
Please check https://github.com/argoproj/argo-cd/issues/11494
I'm waiting for feedback for a PR that would solve this issue

**Comment by greenkiwi** [comment]
We have also been having some issues where having a sync timeout would be very useful.  Changes just get stuck in a syncing state and never progress or fail.  Then the changes to fix the issue aren't able to sync.

**Comment by JavierPorron** [comment]
I'm facing many issues because of this... would be lovely to have a sync timeout

**Comment by rumstead** [comment]
As of Argo CD v2.14, a global sync timeout can be configured for applications. 
https://blog.argoproj.io/argo-cd-v2-14-release-candidate-57a664791e2a

The global sync timeout feature addresses issues where the sync process gets stuck indefinitely, often due to resources failing to start properly and continuously retrying. Now, a global timeout can be configured on the Argo CD Controller to timeout long-running sync operations thanks to [Andrii Korotkov](https://github.com/andrii-korotkov-verkada) (Verkada)!

You can configure the timeout by updating `controller.sync.timeout.seconds` in the argocd-cmd-params-cm configmap.

**Reply by yugaa22** [reply]
> As of Argo CD v2.14, a global sync timeout can be configured for applications. https://blog.argoproj.io/argo-cd-v2-14-release-candidate-57a664791e2a
> 
> The global sync timeout feature addresses issues where the sync process gets stuck indefinitely, often due to resources failing to start properly and continuously retrying. Now, a global timeout can be configured on the Argo CD Controller to timeout long-running sync operations thanks to [Andrii Korotkov](https://github.com/andrii-korotkov-verkada) (Verkada)!
> 
> You can configure the timeout by updating `controller.sync.timeout.seconds` in the argocd-cmd-params-cm configmap.

did any one tried out this?

**Reply by andrii-korotkov-verkada** [reply]
Yes, I've tried and it works. However, if there are multiple retries, I think each gets its own timeout.

**Reply by dlemfh** [reply]
Per my tests, the retry attempts within a single sync operation (set via `spec.syncPolicy.retry`) do not get their own individual timeouts.

The `controller.sync.timeout.seconds` applies to the single entire sync operation **across its multiple retry attempts**.

**Comment by mugdha-adhav** [comment]
Can we also add support for setting sync timeout on application level? Maybe adding a parameter like `.spec.syncPolicy.timeout`.

In our case we cannot use global sync timeout because we deploy more than 100 apps per cluster, and we don't want to risk incomplete syncs.

**Reply by andrii-korotkov-verkada** [reply]
I just set some large value for the global timeout, like 3 hours. It should be more for edge case catching, when syncs are getting stuck. It probably shouldn't be used in a very precise manner. But if there's a demand, I can file an issue to add application-level too. But still, I think the overhead would outweigh the benefits.

**Reply by tberreis** [reply]
I'd like to be able to configure the timeout on application level too. Is there any update on this?
Any chance to get https://argo-cd.readthedocs.io/en/stable/proposals/sync-timeout/#proposal implemented?

**Reply by jstaf** [reply]
Adding a use case for per-application timeout: although most users might prefer ArgoCD syncs to time out after a reasonable delay (5 minutes?), universal sync timeouts can cause serious problems for applications that run database migrations or other Kubernetes Jobs in an ArgoCD PreSync hook. 

We were suddenly having long running database migration jobs killed after just 5 minutes because someone had set `controller.sync.timeout.seconds`. Spent several days tearing my hair out trying to figure why the database migrations were failing, only to find it was a global setting with no way to make exceptions for specific apps that use long running jobs in PreSync/PostSync hooks. Without a way to exclude specific apps from the timeout settings, we can't use this feature.
